File: 065_Permutations.cpp
// Topic: Permutations
// Description: Return all possible orderings of elements.
// Approach: Backtracking swapping or using used array. We'll use in-place swapping.

#include <bits/stdc++.h>
using namespace std;

class Solution
{
   public:
    vector<vector<int>> permute(vector<int>& nums)
    {
        vector<vector<int>> res;
        function<void(int)> dfs = [&](int i)
        {
            if (i == (int)nums.size())
            {
                res.push_back(nums);
                return;
            }
            for (int j = i; j < (int)nums.size(); ++j)
            {
                swap(nums[i], nums[j]);
                dfs(i + 1);
                swap(nums[i], nums[j]);
            }
        };
        dfs(0);
        return res;
    }
};

int main()
{
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    vector<int> a = {1, 2, 3};
    auto res = Solution().permute(a);
    for (auto& v : res)
    {
        for (int i = 0; i < (int)v.size(); ++i)
        {
            if (i)
                cout << " ";
            cout << v[i];
        }
        cout << "\n";
    }
}

/*
Time Complexity: O(n! * n) to generate all permutations
Space Complexity: O(n) recursion depth + output
*/

File: 065_Permutations.py
# Topic: Permutations
# Description: Return all possible orderings of elements.
# Approach: Backtracking swapping or using used array. We'll use in-place swapping.

from typing import List


class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        result = []

        def dfs(start):
            if start == len(nums):
                result.append(nums[:])
                return
            for i in range(start, len(nums)):
                nums[start], nums[i] = nums[i], nums[start]
                dfs(start + 1)
                nums[start], nums[i] = nums[i], nums[start]

        dfs(0)
        return result


if __name__ == "__main__":
    nums = [1, 2, 3]
    result = Solution().permute(nums)
    for v in result:
        print(v)

"""
Time Complexity: O(n! * n) to generate all permutations
Space Complexity: O(n) recursion depth + output
"""
